packages/transactions/src/new-compile-transaction.ts

```diff
@@ -0,0 +1,39 @@
+import { Address } from '@solana/addresses';
+import { SignatureBytes } from '@solana/keys';
+
+import { CompilableTransaction } from './compilable-transaction';
+import { compileMessage } from './message';
+import { getCompiledMessageEncoder } from './serializers/message';
+
+type TypedArrayMutableProperties = 'copyWithin' | 'fill' | 'reverse' | 'set' | 'sort';
+interface ReadonlyUint8Array extends Omit<Uint8Array, TypedArrayMutableProperties> {
+    readonly [n: number]: number;
+}
+
+export type TransactionMessageBytes = ReadonlyUint8Array & { readonly __brand: unique symbol };
+export type OrderedMap<K extends string, V> = Record<K, V>;
+
+export type NewTransaction = Readonly<{
+    messageBytes: TransactionMessageBytes;
+    signatures: OrderedMap<Address, SignatureBytes | null>;
+}>;
+
+export function compileTransactionMessage(transactionMessage: CompilableTransaction): NewTransaction {
+    const compiledMessage = compileMessage(transactionMessage);
+    const messageBytes = getCompiledMessageEncoder().encode(
+        compiledMessage,
+    ) as ReadonlyUint8Array as TransactionMessageBytes;
+
+    const transactionSigners = compiledMessage.staticAccounts.slice(0, compiledMessage.header.numSignerAccounts);
+    const signatures: OrderedMap<Address, SignatureBytes | null> = {};
+    for (const signerAddress of transactionSigners) {
+        signatures[signerAddress] = null;
+    }
+
+    const transaction: NewTransaction = {
+        messageBytes: messageBytes as TransactionMessageBytes,
+        signatures,
```

Comment:
Should we freeze the `signatures` object at this point?

```suggestion
        signatures: Object.freeze(signatures),
```